gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeTableOperations.java

```diff
@@ -0,0 +1,222 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigLakeTableOperations.class);
+
+  private final BigLakeClient client;
+  private final FileIO fileIO;
+  private final TableName tableName;
+
+  BigLakeTableOperations(BigLakeClient client, FileIO fileIO, TableName tableName) {
+    this.client = client;
+    this.fileIO = fileIO;
+    this.tableName = tableName;
+  }
+
+  // The doRefresh method should provide implementation on how to get the metadata location
+  @Override
+  public void doRefresh() {
+    // Must default to null.
+    String metadataLocation = null;
+    try {
+      HiveTableOptions hiveOptions = client.getTable(tableName).getHiveOptions();
+      if (!hiveOptions.containsParameters(METADATA_LOCATION_PROP)) {
+        throw new NoSuchIcebergTableException(
+            "Table %s is not a valid Iceberg table, metadata location not found", tableName());
+      }
+
+      metadataLocation = hiveOptions.getParametersOrThrow(METADATA_LOCATION_PROP);
+    } catch (NoSuchTableException e) {
+      if (currentMetadataLocation() != null) {
+        // Re-throws the exception because the table must exist in this case.
+        throw e;
+      }
+    }
+
+    refreshFromMetadataLocation(metadataLocation);
+  }
+
+  // The doCommit method should provide implementation on how to update with metadata location
+  // atomically
+  @Override
+  public void doCommit(TableMetadata base, TableMetadata metadata) {
+    boolean isNewTable = base == null;
+    String newMetadataLocation = writeNewMetadataIfRequired(isNewTable, metadata);
+
+    CommitStatus commitStatus = CommitStatus.FAILURE;
+    try {
+      if (isNewTable) {
+        createTable(newMetadataLocation, metadata);
+      } else {
+        updateTable(base.metadataFileLocation(), newMetadataLocation, metadata);
+      }
+
+      commitStatus = CommitStatus.SUCCESS;
+    } catch (AlreadyExistsException | CommitFailedException | CommitStateUnknownException e) {
+      throw e;
+    } catch (Throwable e) {
+      commitStatus = checkCommitStatus(newMetadataLocation, metadata);
+      if (commitStatus == CommitStatus.FAILURE) {
+        throw new CommitFailedException(e, "Failed to commit");
+      }
+
+      if (commitStatus == CommitStatus.UNKNOWN) {
+        throw new CommitStateUnknownException(e);
+      }
+    } finally {
+      try {
+        if (commitStatus == CommitStatus.FAILURE) {
+          LOG.warn("Failed to commit updates to table {}", tableName());
+          io().deleteFile(newMetadataLocation);
+        }
+      } catch (RuntimeException e) {
+        LOG.error(
+            "Failed to cleanup metadata file at {} for table {}",
+            newMetadataLocation,
+            tableName(),
+            e);
+      }
+    }
+  }
+
+  @Override
+  public String tableName() {
+    return String.format(
+        "%s.%s.%s", tableName.getCatalog(), tableName.getDatabase(), tableName.getTable());
```

Comment:
This should be the catalog name that was passed to `BigLakeCatalog`, not the catalog ID for BigLake.

Reply:
Done.